yaAGC/agc_engine.c

```diff
@@ -1740,64 +1541,1066 @@ SimulateDV(agc_t *State, uint16_t divisor)
     uint16_t l = c(RegL);
     int i;
 
-    // Assume A contains the sign of the dividend
-    dividend_sign = a & 0100000;
+    // Assume A contains the sign of the dividend
+    dividend_sign = a & 0100000;
+
+    // Negate A if it was positive
+    if (!dividend_sign)
+      a = ~a;
+    // If A is now -0, take the dividend sign from L
+    if (a == 0177777)
+      dividend_sign = l & 0100000;
+    // Negate L if the dividend is negative.
+    if (dividend_sign)
+      l = ~l;
+
+    // Add 40000 to L
+    l = AddSP16(l, 040000);
+    // If this did not cause positive overflow, add one to A
+    if (ValueOverflowed(l) != AGC_P1)
+      a = AddSP16(a, 1);
+    // Initialize the remainder with the current value of A
+    remainder = a;
+
+    // Record the sign of the divisor, and then take its absolute value
+    divisor_sign = divisor & 0100000;
+    if (divisor_sign)
+      divisor = ~divisor;
+    // Initialize the quotient via a WYD on L (L's sign is placed in bits
+    // 16 and 1, and L bits 14-1 are placed in bits 15-2).
+    quotient_sign = l & 0100000;
+    quotient = quotient_sign | ((l & 037777) << 1) | (quotient_sign >> 15);
+
+    for (i = 0; i < 14; i++)
+    {
+        // Shift up the quotient
+        quotient <<= 1;
+        // Perform a WYD on the remainder
+        remainder_sign = remainder & 0100000;
+        remainder = remainder_sign | ((remainder & 037777) << 1);
+        // The sign is only placed in bit 1 if the quotient's new bit 16 is 1
+        if ((quotient & 0100000) == 0)
+          remainder |= (remainder_sign >> 15);
+        // Add the divisor to the remainder
+        sum = AddSP16(remainder, divisor);
+        if (sum & 0100000)
+          {
+            // If the resulting sum has its bit 16 set, OR a 1 onto the
+            // quotient and take the sum as the new remainder
+            quotient |= 1;
+            remainder = sum;
+          }
+    }
+    // Restore the proper quotient sign
+    a = quotient_sign | (quotient & 077777);
+
+    // The final value for A is negated if the dividend sign and the
+    // divisor sign did not match
+    c(RegA) = (dividend_sign != divisor_sign) ? ~a : a;
+    // The final value for L is negated if the dividend was negative
+    c(RegL) = (dividend_sign) ? remainder : ~remainder;
+}
+
+//-----------------------------------------------------------------------------
+// These functions implement scaler timing logic. The scaler is a 33-stage
+// binary counter whose outputs are used to time many things throughout the
+// AGC. Each stage has four outputs -- the inverted and noninverted state for
+// that stage (named, eg., FS09 and FS09/), and two timing pulses of the form
+// FxxA and FxxB. Timing pulse FxxA is emitted when the stage transitions from
+// a 1 to a 0, and timing pulse FxxB is emitted when the stage transitions
+// from a 0 to a 1. The frequency of each stage can be calculated by the
+// forumula (1.024 kHz)/(2^x), where x is the number of the stage.
```

Comment:
Assuming that the comment below is true,

```
// ... Each FS03 count
// therefore happens in half that time, every 39.0625 microseconds. This turns
// out to be (10/3)*MCT.
```

then it must be

```suggestion
// formula (102.4 kHz)/(2^x), where x is the number of the stage.
```

According to https://en.wikipedia.org/wiki/Apollo_Guidance_Computer#Timing the input to the scaler was the crystal oscillator signal of 2.048 MHz divided by 20 (2 x 2 x 5). The Wikipedia article seems to describe Block I (because it mentions only 17 scaler stages), but it must have been the same in Block II.